chromadb/utils/embedding_functions.py

```diff
@@ -31,6 +33,19 @@
 logger = logging.getLogger(__name__)
 
 
+def _retry_call(call):
```

Comment:
Thanks @tazarov!

> A few comments on this:
> 
> * Does it make sense to move this to a separate util module

Yes that makes sense. But then again I don't see it used elsewhere at the moment.
> * Does it make sense to have `_` for a decorator? Seems odd

I agree with you. It was just a leftover from my first commit where I had it as a private function within the class
> * This decorator seems to erase type info.

Yeah I'll fix it.
> * Does it make sense to add configuration options e.g. how many retries, how much to wait, ignored exceptions?

Yes that makes sense. My idea was to have each EmbeddingFunction instance carry its own arguments for `Tenacity.retry` through `self.wait`, which is what I wanted to achieve with the `super().__init__()`. Another thing I can do is avoid the `super()` all together and just let the `call_wrapper_factory` check whether `self` has attribute `wait`. If not, set the decorator to be the default `wait_exponential_jitter()`.
> * The `@retry` decorator seems to throw RetryException. Does it make sense to raise the original exception as it will be better DX

Important catch here. I'll see how tenacity suggest raise the original exception.

Reply:
AFAIK there's no way to implement a decorator like this in python `<= 3.10` without destroying type info. Our auth and telemetry decorators destroy type info and it's something  I would like to fix.

The fact that we can't use a decorator and preserve type info makes me think we shouldn't do it. Instead, we could (in order of my preference off the top of my head):
- Use a [contextmanager](https://docs.python.org/3/library/contextlib.html#contextlib.contextmanager)
- Just use `tenacity` directly wherever we call out to embedding providers.
- Have an explicit function `try_with_retries` or something, which takes retry parameters and the relevant method call.

Reply:
Hey @beggers . I pushed a small update addressing @tazarov's remarks.

Regarding your points:
> AFAIK there's no way to implement a decorator like this in python `<= 3.10` without destroying type info. Our auth and telemetry decorators destroy type info and it's something I would like to fix.
> 
> The fact that we can't use a decorator and preserve type info makes me think we shouldn't do it. Instead, we could (in order of my preference off the top of my head):
> 
> * Use a [contextmanager](https://docs.python.org/3/library/contextlib.html#contextlib.contextmanager)

I'm not familiar with contextlib but I'm looking into it now. I want to ask, IIUC the problem with type-info exists in other places in the repo. Should we push for a solution that solves all of these together?
> * Just use `tenacity` directly wherever we call out to embedding providers.

> * Have an explicit function `try_with_retries` or something, which takes retry parameters and the relevant method call.

I can do both. Let me think about these.